test/Microsoft.AspNet.Mvc.Test/MvcServiceCollectionExtensionsTest.cs

```diff
@@ -79,6 +83,82 @@ public void WithControllersAsServices_ScansControllersFromSpecifiedAssemblies()
             Assert.Equal(ServiceLifetime.Singleton, services[3].Lifetime);
         }
 
+        // Some MVC services can be registered multiple times, for example, 'IConfigureOptions<MvcOptions>' can
+        // be registered by calling 'ConfigureMvc(...)' before the call to 'AddMvc()' in which case the options
+        // configiuration is run in the order they were registered.
+        // For these kind of multi registration service types, we want to make sure that MVC appends its services
+        // to the list i.e it does a 'Add' rather than 'TryAdd' on the ServiceCollection.
+        [Fact]
+        public void MultiRegistrationServiceTypes_AreRegistered_MultipleTimes()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            var multiRegistrationServiceTypes = MutliRegistrationServiceTypes;
+
+            // Act
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+
+            // Assert
+            foreach (var serviceType in multiRegistrationServiceTypes)
+            {
+                AssertServiceCountEquals(services, serviceType, 2);
+            }
+        }
+
+        [Fact]
+        public void SingleRegistrationServiceType_IsNotRegistered_MultipleTimes()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            var multiRegistrationServiceTypes = MutliRegistrationServiceTypes;
+
+            // Act
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+            MvcServiceCollectionExtensions.AddMvcServices(services);
+
+            // Assert
+            var singleRegistrationServiceTypes = services
+                .Where(sd => !multiRegistrationServiceTypes.Contains(sd.ServiceType))
+                .Select(sd => sd.ServiceType);
+
+            foreach (var singleRegistrationType in singleRegistrationServiceTypes)
+            {
+                AssertServiceCountEquals(services, singleRegistrationType, 1);
+            }
+        }
+
+        private IEnumerable<Type> MutliRegistrationServiceTypes
+        {
+            get
+            {
+                return new[]
+                {
+                    typeof(IConfigureOptions<MvcOptions>),
+                    typeof(IConfigureOptions<RazorViewEngineOptions>),
+                    typeof(IActionConstraintProvider),
+                    typeof(IActionDescriptorProvider),
+                    typeof(IActionInvokerProvider),
+                    typeof(IFilterProvider),
+                    typeof(IApiDescriptionProvider)
```

Comment:
The comment in `AddMvcServices()` I requested above should describe what is special about the members of this collection.